yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts

```diff
@@ -1,469 +1,581 @@
+      index: overrides.index ?? BigInt(0),
+      l2BlockNumber: overrides.l2BlockNumber ?? 1,
+      ...overrides,
+    });
+  }
+
+  // Sets up a fresh NoteDataProvider with two scopes and three notes.
+  async function setupProviderWithNotes(storeName: string) {
+    const store = await openTmpStore(storeName);
+    const provider = await NoteDataProvider.create(store);
+
+    await provider.addScope(scope1);
+    await provider.addScope(scope2);
+
+    const note1 = await mkNote({ contractAddress: contractA, storageSlot: slotX, recipient: scope1, index: BigInt(1) });
+    const note2 = await mkNote({ contractAddress: contractA, storageSlot: slotY, recipient: scope1, index: BigInt(2) });
+    const note3 = await mkNote({ contractAddress: contractB, storageSlot: slotX, recipient: scope2, index: BigInt(3) });
+
+    await provider.addNotes([note1, note2], scope1);
+    await provider.addNotes([note3], scope2);
+
+    return { store, provider, note1, note2, note3 };
+  }
+
+  function mkNullifier(note: NoteDao, blockNumber?: number) {
+    return {
+      data: note.siloedNullifier,
+      l2BlockNumber: blockNumber ?? note.l2BlockNumber,
+      l2BlockHash: L2BlockHash.fromString(note.l2BlockHash),
+    };
   }
 
-  // --- Before Each ---
+  describe('NoteDataProvider.create', () => {
+    beforeEach(() => {
+      contractA = AztecAddress.fromString('0xdeadbeef00000000000000000000000000000000000000000000000000000000');
+      contractB = AztecAddress.fromString('0xfeedface00000000000000000000000000000000000000000000000000000000');
+      scope1 = AztecAddress.fromString('0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa');
+      scope2 = AztecAddress.fromString('0xbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb');
+      slotX = Fr.fromString('0x01');
+      slotY = Fr.fromString('0x02');
+    });
+
+    it('creates provider on an empty store and confirms getNotes returns an empty array', async () => {
+      const store = await openTmpStore('note_data_provider_fresh_store');
+      const provider = await NoteDataProvider.create(store);
+
+      const res = await provider.getNotes({ contractAddress: contractA });
+      expect(Array.isArray(res)).toBe(true);
+      expect(res).toHaveLength(0);
+
+      await store.close();
+    });
+
+    it('re-initializes from an existing store and restores previously added notes', async () => {
+      const store = await openTmpStore('note_data_provider_re-init_test');
+
+      // First provider populates the store; second reopens it to verify persistence
+      const provider1 = await NoteDataProvider.create(store);
+
+      await provider1.addScope(scope1);
+      await provider1.addScope(scope2);
+
+      const noteA = await mkNote({ contractAddress: contractA, recipient: scope1, index: BigInt(1) });
+      const noteB = await mkNote({ contractAddress: contractB, recipient: scope2, index: BigInt(2) });
+      await provider1.addNotes([noteA, noteB], AztecAddress.ZERO);
```

Comment:
```suggestion
      const scope = await AztecAddress.random();
      await provider1.addNotes([noteA, noteB], scope);
```
Just a detail but I think it's better to use random value here as zero address made me wonder whether zero scope has some special meaning.

Reply:
I have a weird paranoia about random() even though the chances a collision or failure is astronomically small.   I am adding a constant to make it clear that this is a valid (non-zero) address, and the test will remain deterministic.